File: 문제풀이/2024/052225 백준 KCM Travel 10217 P5.py
from sys import maxsize, stdin, stdout
input = stdin.readline
INF = float('inf')


def solution(airports, budget, flights_count, flights):
    arrival = airports

    dp = [[maxsize]*(budget+1) for _ in range(airports + 1)]
    dp[1][0] = 0
    curr_min = maxsize

    for curr_cost in range(budget+1):
        for airport in range(1, arrival+1):
            if dp[airport][curr_cost] == maxsize:
                continue

            for end, cost, distance in flights[airport]:
                if curr_cost + cost > budget:
                    continue

                dp[end][curr_cost+cost] = \
                    min(dp[end][curr_cost + cost],
                        dp[airport][curr_cost] + distance)

    min_distance = min(dp[arrival])

    return 'Poor KCM' if min_distance == maxsize else str(min_distance)


def main():
    test_cases = int(input())
    answers = []

    for _ in range(test_cases):
        airports, budget, flights_count = map(int, input().split())

        # flights[start] = [(end, cost, distance)]
        flights = [[] for _ in range(airports + 1)]
        for _ in range(flights_count):
            start, end, cost, distance = map(int, input().split())
            flights[start].append((end, cost, distance))

        answer = solution(airports, budget, flights_count, flights)
        answers.append(str(answer))

    stdout.writelines(answers)

    return 0


if __name__ == "__main__":
    main()

File: 문제풀이/2024/052225 백준 KCM Travel 10217 P5.rs
use std::collections::VecDeque;

fn input() -> VecDeque<Vec<usize>> {
    std::io::read_to_string(std::io::stdin())
        .unwrap()
        .trim()
        .split('\n')
        .map(|str| {
            str.trim()
                .split_ascii_whitespace()
                .map(|st| st.parse::<usize>().unwrap())
                .collect::<Vec<usize>>()
        })
        .collect::<VecDeque<Vec<usize>>>()
}

fn solution(airports: usize, budget: usize, flights: Vec<Vec<usize>>) -> String {
    let mut graph = vec![vec![]; airports + 1];

    for line in flights {
        let (from, to, cost, distance) = (line[0], line[1], line[2], line[3]);
        graph[from].push((to, cost, distance));
    }

    // dp[airport][cost] = [distance]
    let mut dp = vec![vec![usize::MAX; budget + 1]; airports + 1];
    dp[1][0] = 0;

    for cost in 0..=budget {
        for airport in 1..=airports {
            if dp[airport][cost] == usize::MAX {
                continue;
            }

            for line in &graph[airport] {
                let (to, flight_cost, flight_distance) = *line;
                let next_cost = cost + flight_cost;
                if next_cost > budget {
                    continue;
                }

                dp[to][next_cost] = dp[to][next_cost].min(dp[airport][cost] + flight_distance);
            }
        }
    }

    match dp[airports]
        .iter()
        .fold(usize::MAX, |curr_min, new| curr_min.min(*new))
    {
        usize::MAX => "Poor KCM".to_string(),
        ret => ret.to_string(),
    }
}

fn main() {
    let mut input = input();

    let test_cases = input.pop_front().unwrap().pop().unwrap();

    let answer = (0..test_cases)
        .map(|_| {
            let line = input.pop_front().unwrap();
            let (airports, budget, flights) = (line[0], line[1], line[2]);

            solution(airports, budget, input.drain(..flights).collect())
        })
        .fold(String::new(), |mut acc, answer| {
            acc.push_str(&answer);
            acc.push('\n');
            acc
        });

    println!("{}", answer);
}
